### ISSUE TITLE
nix-build API for a graphical front-end?

### ISSUE THREAD

**Issue by lukego** [NixOS/nix#1297]
I am interested in making a GUI progress meter for `nix-build`. The basic idea is to summarize what work needs to be done (derivations to build), what is happening at the moment (evaluating/downloading/building), and optionally more details (e.g. click to tail the log.)

(I suppose that this is very similar to the way that Hydra shows evaluations.)

Question is: what would be the most suitable software interface to get the status information about the derivations in real-time? (I am imagining a similar but cleaner alternative to screen-scraping `nix-build` output.)

**Comment by vcunat** [comment]
IIRC dbus was one mentioned for this.

**Comment by expipiplus1** [comment]
At the moment I'm using `NIX_BUILD_HOOK` to intercept and fulfil build requests. I believe @grahamc has a small bash script demonstrating this too.

**Comment by grahamc** [comment]
After learning more things, I determined it was better to read the source code to nix-build (a simple perl script.)

It uses `nix-instantiate` to get a list of drvs to build, and then passes them to `nix-store`. Pretty simple, and you can inspect the drvs for its outputs and watch for them to appear, for example.

**Comment by expipiplus1** [comment]
How does that handle `nix-instantiate` requiring derivations to be built for evaluation (IFD)?

**Comment by taktoa** [comment]
I think the right way to handle this is to modify the Nix sandbox such that it always creates a [ZeroMQ][zeromq] [SUB socket][zeromq-sub] in a specified location, and then the builder can write messages to it in some specified format ([Protobuf][protobuf] or [Cap'n Proto][capnproto] or [CBOR][cbor] or just plain old JSON text would be decent for this as long as the format has a schema; [JSON Schema][jsonschema] can be used if the format is CBOR or JSON text).

Note that it is [possible](http://zeromq.org/bindings:bash) to use ZeroMQ from Bash, so this wouldn't even cause any bootstrapping problems.

ZeroMQ is simpler than DBus and more portable than raw Unix domain sockets or FIFOs, so I think it is probably the right way to do this kind of IPC.

I'll try hacking something together, and if I can get it working I'll explore what kinds of data we might want in the messages coming out of a build process.

[zeromq]:     http://zeromq.org
[zeromq-sub]: http://zguide.zeromq.org/page:all#Getting-the-Message-Out
[protobuf]:   https://github.com/google/protobuf
[capnproto]:  https://capnproto.org
[cbor]:       http://cbor.io
[jsonschema]: http://json-schema.org

**Comment by copumpkin** [comment]
Also relevant (especially to what @taktoa was saying): https://github.com/NixOS/nix/issues/896

**Comment by expipiplus1** [comment]
Making a new store implementation is another way to intercept calls to the nix store. I've got a minimal example here https://github.com/expipiplus1/dummy-store

**Comment by taktoa** [comment]
I'll move my discussion to #896

**Comment by lukego** [comment]
Just now I am pretty happy with the approach that @grahamc suggested i.e. use `nix-instantiate` to work out which derivations need to be built and then use `nix-store` to poll for their progress while one or more builds are running.

**Comment by vcunat** [comment]
I'm using `ls -ld /tmp/nix-*` to see what's still being built :-)

**Comment by lukego** [comment]
Is there a new answer to this question in light of new progress reporting functionality? https://github.com/NixOS/nix/issues/896#issuecomment-323717243

I am interested in adding progress tracking to a graphical frontend to `nix-build` right now and I am wondering what approach to take.

**Comment by edolstra** [comment]
Yes. The cleanest way to do this would be to subclass `Logger` with a class that keeps track of active activities and updates the GUI appropriately, similar to `ProgressBar`'s single-line TUI.

**Comment by lukego** [comment]
Just at first glance it seems like the simplest solution for me would be for Nix to write all the raw progress events into a JSON file and the GUI could stream this for updates.

Is there anybody already working on something like that? Just thinking that it could serve as a common interface for frontend tools that are not running inside the Nix C++ code.

**Comment by edolstra** [comment]
Yes, if you need to have the GUI in a different process, a `JSONLogger` would be the way to go, I think.

**Comment by lukego** [comment]
Thanks @edolstra. I'm looking at hacking this now. I'm not sure if I will succeed though because I don't know C++ :).

**Comment by lukego** [comment]
I started writing `JSONLogger` but now I noticed `BuilderLogger` that seems to already dump the events at JSON. So `JSONLogger` will be almost identical, right? Just will make sure each output line is well-formed JSON and will hook in more like `progress-bar.cc` does?

**Comment by edolstra** [comment]
Yeah, it should be possible to reuse `BuilderLogger` for this.

**Comment by tomberek** [comment]
Closing as stale. Re-open if needed.
